Electron/Core/RoveProtocol/Rovecomm.ts

```diff
@@ -1,85 +1,273 @@
+   * Note: even if dataLength is only 1, this returns an array of one item.
+   */
+
+  let readBytes: (i: number) => number
+
+  switch (dataType) {
+    case DataTypes.INT8_T:
+      readBytes = data.readInt8.bind(data)
+      break
+    case DataTypes.UINT8_T:
+      readBytes = data.readUInt8.bind(data)
+      break
+    case DataTypes.INT16_T:
+      readBytes = data.readInt16BE.bind(data)
+      break
+    case DataTypes.UINT16_T:
+      readBytes = data.readUInt16BE.bind(data)
+      break
+    case DataTypes.INT32_T:
+      readBytes = data.readInt32BE.bind(data)
+      break
+    case DataTypes.UINT32_T:
+      readBytes = data.readUInt32BE.bind(data)
+      break
+    case DataTypes.FLOAT_T:
+      readBytes = data.readFloatBE.bind(data)
+      break
+    default:
+      return []
+  }
+
+  const retArray = []
+  let offset: number
+  for (let i = 0; i < dataLength; i += 1) {
+    offset = i * dataSizes[dataType]
+    retArray.push(readBytes(offset))
+  }
+  return retArray
+}
+
+function parse(packet: Buffer): void {
+  /*
+   * Parse takes in a packet buffer and will call decodePacket and emit
+   *  the rovecomm event with the proper dataId and that typed data
+   *
+   *  RoveComm Header Format:
+   *
+   *   0                   1                   2                   3
+   *   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   *  |    Version    |            Data Id            |  Data Length  |
+   *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   *  |   Data Type   |                Data (Variable)                |
+   *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   *
+   *  Note: the size of Data is dataSizes[DataType] * dataSizes bytes
+   */
+
+  const VersionNumber = 2
+
+  const version = packet.readUInt8(0)
+  const dataId = packet.readUInt16BE(1)
+  const dataLength = packet.readUInt8(3)
+  const dataType = packet.readUInt8(4)
+
+  const rawdata = packet.slice(5)
+  let data: number[]
+
+  if (version === VersionNumber) {
+    data = decodePacket(dataType, dataLength, rawdata)
+
+    let dataIdStr = "null"
+    // Here we loop through all of the Boards in the manifest,
+    // looking specifically if this dataId is a known Telemetry of the board
+    for (let i = 0; i < DATAID.length; i++) {
+      if (dataId in DATAID[i].Telemetry) {
+        dataIdStr = DATAID[i].Telemetry[dataId]
+      }
+    }
+
+    // rovecomm depends on parse, and parse depends on rovecomm. Since parse is defined with the
+    // function keyword, this isn't a problem, but we don't want to disable this use before definition
+    // error for the whole file or project since it can be risky
+
+    // First emit is for the dataIdString (like DriveSpeeds)
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    rovecomm.emit(dataIdStr, data)
```

Comment:
Do you want to handle telemetry we can't find in the manifest? Or is a null dataIdStr event good enough?

Reply:
With the current structure, the emit with a specific string will be used for easy consumption of messages by down-stream components, so leaving it as null will just prevent any component from accessing the data. It will still be logged through the "all" emit just after.

It would probably be best to have an additional warning in the event that a "null" emit occurs to provide real-time feedback as to packet definitions varying between the board implementation and the basestation implementation though.

Reply:
Emitting this null event will log the data and the dataId to the Console. The hope is that this will be enough information for someone at basestation to diagnose this issue and fix the board code or add the dataId to the manifest.